### PR TITLE
Add MXNet Gluon model functionality.

### PR DESCRIPTION
**Pull request by meissnereric**
We have some models that are built using MXNet's Gluon interface (http://gluon.mxnet.io/), so I added another model class to handle that type of network.

Also, there is a bug in the mxnet codebase with using the mx.symbol.softmax_cross_entropy and it hasn't worked in a while so I fixed that in the existing MXNet code as well.

Just contributing back if others want to use MXNet models :) Feedback welcome.

Thanks,
Eric

### CODE REVIEW

**Comment by jonasrauber** [comment]
Hi @meissnereric, thanks a lot for your pull-request. It's great to have support for Gluon models! I am aware of the `softmax_cross_entropy` bug in MXNet; just last week I wanted to finally upgrade our tests to the latest MXNet and ran into it again: https://github.com/apache/incubator-mxnet/issues/6874 so I really appreciate your workaround.

I am looking forward to merging this. Before that, we need to fix the code formatting and update the tests.

Here are some code-formatting issues that need to be fixed:
```
flake8 --ignore E402,E741 .
./foolbox/models/mxnet_gluon.py:77:14: E225 missing whitespace around operator
./foolbox/models/mxnet_gluon.py:84:80: E501 line too long (97 > 79 characters)
./foolbox/models/mxnet_gluon.py:89:14: E225 missing whitespace around operator
./foolbox/models/mxnet.py:7:1: E302 expected 2 blank lines, found 1
./foolbox/models/mxnet.py:66:80: E501 line too long (89 > 79 characters)
./foolbox/models/mxnet.py:68:9: E265 block comment should start with '# '
```

Also, you made some changes to adversarial.py. Could you remove them from the pull-request?

Finally, it would be great if you could upgrade MXNet in the travis config to the latest MXNet version so that we can see if your workaround passes the tests; see here: https://github.com/bethgelab/foolbox/pull/133/files

It would be great if you could to these changes, otherwise I can also do them myself as soon as I find time.

**Comment by meissnereric** [comment]
Cool, I fixed up the formatting, removed the adversarial.py change, and upgraded the travis file.

Thanks,
Eric

File: foolbox/models/mxnet.py [hunk 1]
```diff
@@ -63,7 +63,11 @@ def __init__(
         label = mx.symbol.Variable('label')
         self._label_sym = label
 
-        loss = mx.symbol.softmax_cross_entropy(logits, label)
+        log_softmax = mx.sym.log_softmax(logits)
```
Comment (jonasrauber) [inline foolbox/models/mxnet.py]: can you add a comment pointing out that this is a workaround for https://github.com/apache/incubator-mxnet/issues/6874

File: foolbox/models/mxnet_gluon.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+from __future__ import absolute_import
+
+from .base import DifferentiableModel
+
+import numpy as np
+
+
+class MXNetGluonModel(DifferentiableModel):
+    """Creates a :class:`Model` instance from existing `MXNet` symbols and weights.
```
Comment (jonasrauber) [inline foolbox/models/mxnet_gluon.py]: can you update the documentation and the parameter list

File: foolbox/models/mxnet_gluon.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        image = self._process_input(image)
+        label = mx.nd.array([label])
+        data_array = mx.nd.array(image, ctx=self._device)
+        data_array.attach_grad()
+        with mx.autograd.record(train_mode=False):
+            L = self._block(data_array)
+            loss = mx.nd.softmax_cross_entropy(L, label)
+            loss.backward()
+        return np.squeeze(L.asnumpy(), axis=0),
+        self._process_gradient(data_array.grad.asnumpy())
```
Comment (jonasrauber) [inline foolbox/models/mxnet_gluon.py]: indentation missing

File: foolbox/models/mxnet_gluon.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    def batch_predictions(self, image):
+        import mxnet as mx
+        image = self._process_input(image)
+        data_array = mx.nd.array(image, ctx=self._device)
+        data_array.attach_grad()
+        with mx.autograd.record(train_mode=False):
+            L = self._block(data_array)
+        return np.squeeze(L.asnumpy(), axis=0)
+
+    def predictions(self, image):
```
Comment (jonasrauber) [inline foolbox/models/mxnet_gluon.py]: no need to implement `predictions` if `batch_predictions` is implemented (the base class will handle this)

File: foolbox/models/mxnet_gluon.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            channel_axis=channel_axis,
+            preprocessing=preprocessing)
+
+        self._device = ctx
+        self._block = block
+
+    def num_classes(self):
+        return self._num_classes
+
+    def batch_predictions(self, image):
```
Comment (jonasrauber) [inline foolbox/models/mxnet_gluon.py]: batch_predictions gets a batch of inputs and should be able to calculate predictions for all of them; this implementation only works for single inputs

**Comment by jonasrauber** [comment]
Thanks Eric!
I reviewed the code and listed some changes that are still necessary. In the end, we also need a small `mean_brightness_net` example for the tests, similar to `foolbox/tests/test_models_mxnet.py`.

**Comment by meissnereric** [comment]
Any updates?

File: foolbox/models/mxnet_gluon.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
+from __future__ import absolute_import
+
+from .base import DifferentiableModel
+
+import numpy as np
+
+
+class MXNetGluonModel(DifferentiableModel):
+    """Creates a :class:`Model` instance from existing `MXNet` symbols and weights.
```
Comment (jonasrauber) [inline foolbox/models/mxnet_gluon.py]: This line of the documentation still needs to be updated

**Comment by jonasrauber** [comment]
Thanks a lot, @meissnereric! Looking forward to merging this as soon as that last small change in the documentation of the `MXNetGluonModel` documentation is fixed.

**Comment by jonasrauber** [comment]
Great work! Thanks a lot!

**Pull request merged**